### ISSUE TITLE
feature request: add filter

### ISSUE THREAD

**Issue by thedino-girl** [JabRef/jabref#10156]
Hi

I would like to request the following filters for JabRef:

- author filter
- year filter
- tag filter: each entry can be assigned a tag on which you can apply a filter

You can see this feature in the Mendeley Reference Manager:

https://www.youtube.com/watch?v=OzFHGFnAM2Q&t=527s

**Comment by thedino-girl** [comment]
Maybe the filter can be configured so you can choose what fields are filtered.

**Comment by Siedlerchr** [comment]
HI, this is already possible with groups. Basically, you can use groups for this.
You can create a group for filtering by keyword for free search expression.

https://docs.jabref.org/finding-sorting-and-cleaning-entries/groups#types-of-groups

@ThiloteE  Can probably say more about this

**Comment by thedino-girl** [comment]
But the goal of this feature is to have a list of all authors, years, ... from which the user can choose to quickly filter the database without having to create a group. See the youtube video.

I'm also interested in an example of the free search expression (for example to filter on author) because I don't understand it by reading the documentation.

**Comment by ThiloteE** [comment]
Mendeley filters refer to what other programms often would call "tags", "labels" or "keywords".
Also relevant related issues: https://github.com/JabRef/jabref/issues/8739 and https://github.com/JabRef/jabref/issues/9203.

Yes, it would be nice to have a better representation of tags, filters, labels or keywords in the JabRef UI. Though, most (not all) of what those offer can already be reproduced with existing features of JabRef. The things that cannot be done are mostly just UI things. Meaning, keywords are not displayed in the maintable or somewhere where users can easily click on them to trigger a selected view of certain entries, apart from using the groups feature.

What we have:

1. Groups with searching for keyword
2. Groups with free search expression
3. Groups with specified keywords
4. A "Keywords" field in the "general" tab of the entry editor / or any other custom field, if you like! --> show the "keywords" field in the main table via `options / file > preferences > entry table`. Unfortunately, it is currently not possible to click on those words to selectively show only entries with these words (e.g. like github's labels). Not sure, if Javafx offers something like this.
5. "Search" for (key-) words (and fields, field content, citekeys and entrytypes) within or across groups or within your whole database (within the "all entries" group) using the search bar. In the most incidents the normal search is enough, but if not, then you can always go the advanced route and come up with a regular expression.

Now, if you feel like you want to "save" your search (which would be again something like a label, tag, filter or keyword, then you basically have two choices: A) create a group or B) write the "keyword" directly into one of the fields in your library, which then can be found by the search or groups feature.

Storing your data directly in the library is a great way to ensure colleagues will have access to your keyword once you decide to share your library. There is no further need to also share JabRef settings. I suspect, if JabRef were to implement something like a unique keywords feature, then the easiest approach would be to just come up with a custom field called "labels" or something and then to make this field more prominent and accessible in the UI. There already is the keywords field, but those are usually filled with official data inherent to your entries (e.g. fetched from DOI or ISBN), whereas your own unique labels should probably be put into another (custom) field.

Just a short reminder about JabRef's library syntax (which follows bibtex/biblatex), which you can see in the "bibtex source tab", so when I say "field", I mean part of the following:

```
@Entrytype{citekey,
field = {field content},
}
```

**Comment by ThiloteE** [comment]
While i think the author filter or tag filter are covered by what I wrote above, this is not the case for date ranges.
Enhancing the search or groups feature (or both) with the possibility to search for a range of the "year" or "date" would indeed be nice! E.g. find all entries with data between 2000 and 2012.

I guess it's possible with java. See here:

- https://www.baeldung.com/java-between-dates
- https://www.baeldung.com/java-iterate-date-range

While it's technically possible to come up with some regex, this seems to be highly inefficient and maybe come with very mediocre performance as this source claims: https://stackoverflow.com/questions/6839987/regular-expression-for-a-date-range

**Comment by thedino-girl** [comment]
I think there should be a "tag" field for each entry, just as in Mendeley. 
And this tag should be possible to be filtered, completely independent from the groups with a filter button as in Mendeley.

- The reason for this is that this tag filter is not permanent and should not be stored as a group or anywhere in the bibtex file: the user wants to see all entries of a certain year or with a certain author, ... but does not want to store this filter.

- the filter should also show all possible values in all other fields that already exist (all fields except the tag field): 

1. for year: all years in the database
2. for author: all different authors (separated by "," or AND) in the author field. 
3. ...

So this is also not possible with the groups. You must inspect the database yourself to know what values to enter to search for.
If the filters show list of possible values, this is much more user friendly. You can quickly filter the database and remove the filter again.

So the tag filter should be possible to apply immediately by choosing list of possible values, and not stored in the bibtex file.
Also filters are gone when JabRef is closed.
This was my idea for a filter. 

Is there interest to make this?

**Comment by ThiloteE** [comment]
Since I do not want to write out "tags / filters / keywords / labels", I will simply use the word "keywords" for now.

I think what you want is slightly at odds with each other. At one point you want tags not permanent, but then later you want them permanent.

What JabRef has:

### Not permanent:
   - [x] all your needs are covered by using "search"!
   
### Permanent:

#### A) Data is stored in your library:
   - [x] JabRef stores data in your library (the .bib file). This includes keywords.
      ![image](https://github.com/JabRef/jabref/assets/73715071/34719260-d626-4cdb-b88b-a9692dcbcfea)
   - [x] being part of a group is stored in your entry.
      ![image](https://github.com/JabRef/jabref/assets/73715071/80b6a83b-ceea-4b60-8fa1-c0aa71bc7ffe)
      - [ ] but this stored data does not automatically create new groups in the UI, if you import the library file. See https://github.com/JabRef/jabref/issues/8836
   - [x] The entry table / main table shows you all the possible values and you can sort and scroll through this table to find specific keywords.
   - [x] JabRef searches through your library and finds "keywords", which are stored in your library. Note, these keywords are not limited to what is stored in the keyword field, but can be any field or field content, including authors or year. JabRef will store the search in a separate place in your library file (when you open the file with a text editor, the location can be found via `@Comment{jabref-meta: grouping:`. These tags are then shown separately in a different place in JabRef's UI
     - [x] this is basically "groups with [specified keywords](https://docs.jabref.org/finding-sorting-and-cleaning-entries/groups#specified-keywords)" and all other groups related features.
     - [x] this is also "Manage keywords" from `edit > manage keywords`.
         - [ ] also show field content from other fields in "Manage keywords", other than from the keywords field.
     
#### B) Data is NOT stored in your library, but somewhere else on your computer / remotely.
  
  - [ ] JabRef searches through your library, comes up with tags, which are really not stored in your library, but still shown separately in a different place in the UI
  
Have you actually tried those features, I have pointed you towards?
The next big thing on the agenda is a rework of JabRef's search engine, so it might take a while until core developers have a look at this feature request, especially since it is possible to work around most of the issues you describe with groups and search. If anybody is willing to work on keywords, sure, go ahead :-) 👍

**Comment by thedino-girl** [comment]
Yes I understand but currently it's not really user friendly to work with keywords :)
It takes a lot of time before you understand what is where and how it works.

And you use keywords for "keywords" field that is automatically filled and then yuo use another "keywords" for the screenshot above. Why you don't call it tags in the last case?

![afbeelding](https://github.com/JabRef/jabref/assets/141878924/3f8ea275-3c78-4fee-9136-c8b34b665a6b)

**Comment by ThiloteE** [comment]
> We can currently not assign a user tag we choose ourselves in a separate field?

You can, if you use the groups feature.

**Comment by thedino-girl** [comment]
> e.g. by creating your own custom field "tags" and creating a group using following regex: `field = .+` ![image](https://user-images.githubusercontent.com/73715071/260159123-2713d43b-3645-417a-ad97-707c448faf03.png)
> 
> `.` means: any character `+` means: one or more times

Where is the correct window to make custom fields?

What is the difference between `.+` and `.*` because they both seem to work.

**Comment by ThiloteE** [comment]
Both seem to do the same thing, according to https://docs.jabref.org/finding-sorting-and-cleaning-entries/search#greedy-quantifiers. The documentation originates from https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/regex/Pattern.html

**Comment by ThiloteE** [comment]
You can declare some via the entry types menu.

in JabRef 5.9 go to `options > custom entry types`.
in JabRef 5.10 development version go to `File > preferences > entry types`.

You also can just open the entry editor, go to the biblatex source tab and type them in.
For example:

![image](https://github.com/JabRef/jabref/assets/73715071/577bdf92-3a6f-4a2a-a95a-e24126ad936e)

**Comment by Siedlerchr** [comment]
The regex quantifier: .+ means: 1 or more, .* zero or more

**Comment by ThiloteE** [comment]
Ah, you are right. My bad.

**Comment by thedino-girl** [comment]
> The regex quantifier: .+ means: 1 or more, .* zero or more

Thanks! So there is no way to write an empty string like (where empty is something that works in jabref)?
`tag = empty`

We must use the `.` and use `!=`?
